libcnb/examples/example-02-ruby-sample/src/main.rs

```diff
@@ -36,9 +35,14 @@ impl Buildpack for RubyBuildpack {
     fn build(&self, context: BuildContext<Self>) -> libcnb::Result<BuildResult, Self::Error> {
         println!("---> Ruby Buildpack");
 
-        let ruby_env = execute_layer_lifecycle("ruby", RubyLayerLifecycle, &context)?;
+        let ruby_layer = context.handle_layer("ruby", RubyLayer)?;
```

Comment:
Traits can't have fields, but it could be a method on the `Layer` trait. Having it as a method has some drawbacks as the value could change between method calls. We can avoid this issue by just calling it once, but it's a potential problem. We have something similar with `Layer::types`.

The real reason this was designed the way it was is that `Layer` implementations are meant to be re-usable for different layers on disk. This might at first sound weird as we tend to think about layers as something very specific. But we can totally have, for example, a generic `StaticFileLayer` in a shared library, that writes a file to a layer and puts in on the path:

```rust
struct StaticFileLayer<B: Buildpack> {
    file_name: String,
    file_contents: String,
    error_mapper: Box<dyn Fn(std::io::Error) -> B::Error>,
}

impl<B: Buildpack> Layer for StaticFileLayer<B> {
    type Buildpack = B;
    type Metadata = GenericMetadata;

    fn types(&self) -> LayerTypes {
        LayerTypes {
            cache: false,
            launch: true,
            build: true,
        }
    }

    fn create(
        &self,
        _: &BuildContext<Self::Buildpack>,
        layer_path: &Path,
    ) -> Result<LayerResult<Self::Metadata>, <Self::Buildpack as Buildpack>::Error> {
        std::fs::write(
            &layer_path.join("bin").join(&self.file_name),
            &self.file_contents,
        )
        .and_then(|_| LayerResultBuilder::new(GenericMetadata::default()).build())
        .map_err(&self.error_mapper)
    }
}
```

A buildpack that then wants to use this shared `Layer` implementation can then have this in it's `build` method:

```rust
context.handle_layer("runner", StaticFileLayer<MyBuildpack> {
        file_name: String::from("run.sh"),
        file_contents: String::from(!include_str!("../opt/run.sh")),
        error_mapper: Box::new(MyBuildpackError::RunShIoError),
    })?;
```


The above would be harder to do if the `Layer` would have a fixed layer name.

Reply:
Seems good